### ISSUE TITLE
certain nexus devices don't return IP address table

### ISSUE THREAD

**Issue by ajw1980** [netdisco/snmp-info#346]
It would seem that newer versions of nexus switches don't return the IP address table correctly:

```
~/bin/netdisco-do show -d xxx -e ip_table -DI
[2890] 2019-06-25 15:04:09  info App::Netdisco version 2.042010 loaded.
[2890] 2019-06-25 15:04:10  info show: [10.30.0.185]/ip_table started at Tue Jun 25 10:04:10 2019
[2890] 2019-06-25 15:04:10 debug show: running with timeout 600s
[2890] 2019-06-25 15:04:10 debug => running workers for phase: check
[2890] 2019-06-25 15:04:10 debug -> run worker check/_base_/0
[2890] 2019-06-25 15:04:10 debug Show is able to run
[2890] 2019-06-25 15:04:10 debug => running workers for phase: main
[2890] 2019-06-25 15:04:10 debug -> run worker main/_base_/100
[2890] 2019-06-25 15:04:10 debug snmp reader cache warm: [10.30.0.185]
[2890] 2019-06-25 15:04:10 debug [10.30.0.185:161] try_connect with ver: 2, class: SNMP::Info::Layer3::Nexus, comm: <hidden>
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
SNMP::Info::_global id : SNMPv2-MIB::sysObjectID.0 : .1.3.6.1.2.1.1.2.0
SNMP::Info 3.68
SNMP::Info::device_type() layers:01000110 id:9 sysDescr:"Cisco NX-OS(tm) nxos.7.0.3.I7.2.bin, Software (nxos), Version 7.0(3)I7(2), RELEASE SOFTWARE Copyright (c) 2002-2016 by Cisco Systems, Inc. Compiled 11/22/2017 12:00:00"
SNMP::Info::_load_attr old_ip_table : IP-MIB::ipAdEntAddr : .1.3.6.1.2.1.4.20.1.1
SNMP::Info::_load_attr new_ip_index : IP-MIB::ipAddressIfIndex : .1.3.6.1.2.1.4.34.1.3
SNMP::Info::_load_attr new_ip_type : IP-MIB::ipAddressType : .1.3.6.1.2.1.4.34.1.4
SNMP::Info::_load_attr udpLocalAddress : UDP-MIB::udpLocalAddress : .1.3.6.1.2.1.7.5.1.1
\ {}
[2890] 2019-06-25 15:04:10  info show: finished at Tue Jun 25 10:04:10 2019
[2890] 2019-06-25 15:04:10  info show: status done: Showed ip_table response from 10.30.0.185
```
Older version looks ok:
```
~/bin/netdisco-do show -d xxx -e ip_table -DI
[2870] 2019-06-25 15:04:01  info App::Netdisco version 2.042010 loaded.
[2870] 2019-06-25 15:04:01  info show: [10.29.0.32]/ip_table started at Tue Jun 25 10:04:01 2019
[2870] 2019-06-25 15:04:01 debug show: running with timeout 600s
[2870] 2019-06-25 15:04:01 debug => running workers for phase: check
[2870] 2019-06-25 15:04:01 debug -> run worker check/_base_/0
[2870] 2019-06-25 15:04:01 debug Show is able to run
[2870] 2019-06-25 15:04:01 debug => running workers for phase: main
[2870] 2019-06-25 15:04:01 debug -> run worker main/_base_/100
[2870] 2019-06-25 15:04:01 debug snmp reader cache warm: [10.29.0.32]
[2870] 2019-06-25 15:04:01 debug [10.29.0.32:161] try_connect with ver: 2, class: SNMP::Info::Layer3::Nexus, comm: <hidden>
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
SNMP::Info::_global id : SNMPv2-MIB::sysObjectID.0 : .1.3.6.1.2.1.1.2.0
SNMP::Info 3.68
SNMP::Info::device_type() layers:01000110 id:9 sysDescr:"Cisco NX-OS(tm) nxos.7.0.3.I4.5.bin, Software (nxos), Version 7.0(3)I4(5), RELEASE SOFTWARE Copyright (c) 2002-2013 by Cisco Systems, Inc. Compiled 12/23/2016 15:00:00"
SNMP::Info::_load_attr old_ip_table : IP-MIB::ipAdEntAddr : .1.3.6.1.2.1.4.20.1.1
SNMP::Info::_load_attr udpLocalAddress : UDP-MIB::udpLocalAddress : .1.3.6.1.2.1.7.5.1.1
\ {
    10.29.103.1     "10.29.103.1",
    10.29.104.100   "10.29.104.100",
    10.29.250.1     "10.29.250.1",
    10.29.254.9     "10.29.254.9",
    192.168.30.1    "192.168.30.1",
    192.168.31.1    "192.168.31.1",
    192.168.32.1    "192.168.32.1",
    192.168.33.1    "192.168.33.1",
    192.168.35.1    "192.168.35.1",
    192.168.36.1    "192.168.36.1",
    192.168.37.1    "192.168.37.1"
}
[2870] 2019-06-25 15:04:02  info show: finished at Tue Jun 25 10:04:02 2019
[2870] 2019-06-25 15:04:02  info show: status done: Showed ip_table response from 10.29.0.32
```

**Comment by ajw1980** [comment]
Ok. Nothing returned on SNMP for ipAddrTable for those devices:
IP-MIB::ipAddrTable = No Such Object available on this agent at this OID

**Comment by inphobia** [comment]
strange, mine work just fine. 9332pq running 7.0.3.I7.5a.

```
netdisco-do show -d 10.1.100.250 -e ip_table -DI
[22618] 2019-06-25 20:20:23  info App::Netdisco version 2.042010 loaded.
[22618] 2019-06-25 20:20:23  info show: [10.1.100.250]/ip_table started at Tue Jun 25 22:20:23 2019
[22618] 2019-06-25 20:20:23 debug show: running with timeout 600s
[22618] 2019-06-25 20:20:23 debug => running workers for phase: check
[22618] 2019-06-25 20:20:23 debug -> run worker check/_base_/0
[22618] 2019-06-25 20:20:23 debug Show is able to run
[22618] 2019-06-25 20:20:23 debug => running workers for phase: main
[22618] 2019-06-25 20:20:23 debug -> run worker main/_base_/100
[22618] 2019-06-25 20:20:23 debug snmp reader cache warm: [10.1.100.250]
[22618] 2019-06-25 20:20:23 debug [10.1.100.250:161] try_connect with ver: 3, class: SNMP::Info::Layer3::Nexus, comm: <hidden>
SNMP::Info::_global uptime : SNMPv2-MIB::sysUpTime.0 : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
SNMP::Info::_global id : SNMPv2-MIB::sysObjectID.0 : .1.3.6.1.2.1.1.2.0
SNMP::Info 3.68
SNMP::Info::device_type() layers:01000110 id:9 sysDescr:"Cisco NX-OS(tm) nxos.7.0.3.I7.5a.bin, Software (nxos), Version 7.0(3)I7(5a), RELEASE SOFTWARE Copyright (c) 2002-2018 by Cisco Systems, Inc. Compiled 10/12/2018 18:00:00"
SNMP::Info::_load_attr old_ip_table : IP-MIB::ipAdEntAddr : .1.3.6.1.2.1.4.20.1.1
SNMP::Info::_load_attr udpLocalAddress : UDP-MIB::udpLocalAddress : .1.3.6.1.2.1.7.5.1.1
\ {
    10.1.30.1       "10.1.30.1",
    10.1.100.250    "10.1.100.250",
...
```

i am running snmpv3 (perhaps your default snmp view changed?), also note that vrfs are a work in progress (non defaults vrfs don't yet work for me).

the thing that stands out to me is that your non-working device goes through 
```
SNMP::Info::_load_attr old_ip_table : IP-MIB::ipAdEntAddr : .1.3.6.1.2.1.4.20.1.1
SNMP::Info::_load_attr new_ip_index : IP-MIB::ipAddressIfIndex : .1.3.6.1.2.1.4.34.1.3
SNMP::Info::_load_attr new_ip_type : IP-MIB::ipAddressType : .1.3.6.1.2.1.4.34.1.4
SNMP::Info::_load_attr udpLocalAddress : UDP-MIB::udpLocalAddress : .1.3.6.1.2.1.7.5.1.1
```
while the working one is happy with just ipAdEntAddr & udpLocalAddress the non working one seems to be doing more calls for some reason.


can you provide the following output of the non-working one?
```
netdisco-do show -d 10.30.0.185 -DI -e ::ip_table
netdisco-do show -d 10.30.0.185 -DI -e udpLocalAddress
netdisco-do show -d 10.30.0.185 -DI -e _local_addr 
```

**Comment by ajw1980** [comment]
Looking at that switch specific switch a little more, it only has mgmt0 enabled. It does not return anything on .1.3.6.1.2.1.4.20:
```
snmpwalk -v2c -c public xxx .1.3.6.1.2.1.4.20
IP-MIB::ipAddrTable = No Such Object available on this agent at this OID
```
Maybe the nexus doesn't report anything if only mgmt0 has an address?

I also have another nexus switch that does report addresses in the ip_table, but nothing gets in to netdisco. What do I need to send for that?

**Comment by inphobia** [comment]
hm, don't really have a nexus here with only a management config atm. what does `sh snmp internal oids registered | inc iso.3.6.1.2.1.4.20` on the switch show?

> but nothing gets in to netdisco. What do I need to send for that?

could you elaborate on this, as in what is "nothing gets into netdisco" mean.

a netdisco-do discover with options -DIIS would help but could contain some sensitive info.

**Comment by ajw1980** [comment]
Ok. Nothing in the default vrf and the mgmt0 port is on its own vrf.

```
# sh snmp internal oids registered | inc iso.3.6.1.2.1.4.20
02   iso.3.6.1.2.1.4.19 - iso.3.6.1.2.1.4.20.1.1
02   iso.3.6.1.2.1.4.20.1.1 - iso.3.6.1.2.1.4.20.1.2
02   iso.3.6.1.2.1.4.20.1.2 - iso.3.6.1.2.1.4.20.1.3
02   iso.3.6.1.2.1.4.20.1.3 - iso.3.6.1.2.1.4.20.1.4
02   iso.3.6.1.2.1.4.20.1.4 - iso.3.6.1.2.1.4.20.1.5
02   iso.3.6.1.2.1.4.20.1.5 - iso.3.6.1.2.1.4.21.1.1
# show ip interface

IP Interface Status for VRF "default"
# show ip interface vrf all

IP Interface Status for VRF "default"

IP Interface Status for VRF "management"
mgmt0, Interface status: protocol-up/link-up/admin-up, iod: 3,
  IP address: 10.30.0.185, IP subnet: 10.30.0.0/23 route-preference: 0, tag: 0
  IP broadcast address: 255.255.255.255
  IP multicast groups locally joined: none
  IP MTU: 1500 bytes (using link MTU)
  IP primary address route-preference: 0, tag: 0
  IP proxy ARP : disabled
  IP Local Proxy ARP : disabled
  IP multicast routing: disabled
  IP icmp redirects: enabled
  IP directed-broadcast: disabled
  IP Forwarding: disabled
  IP icmp unreachables (except port): disabled
  IP icmp port-unreachable: enabled
  IP unicast reverse path forwarding: none
  IP load sharing: none
  IP interface statistics last reset: never
  IP interface software stats: (sent/received/forwarded/originated/consumed)
    Unicast packets    : 78041499/77705136/0/78041499/155410272
    Unicast bytes      : 25659522996/9594528677/0/25659522996/19188610282
    Multicast packets  : 0/38439488/0/0/0
    Multicast bytes    : 0/3451580580/0/0/0
    Broadcast packets  : 0/96958703/0/0/0
    Broadcast bytes    : 0/8787468936/0/0/0
    Labeled packets    : 0/0/0/0/0
    Labeled bytes      : 0/0/0/0/0
  WCCP Redirect outbound: disabled
  WCCP Redirect inbound: disabled
  WCCP Redirect exclude: disabled
```

For the other switch, by "nothing gets in to netdisco", I mean none of the IP addresses get attached to the device. The rest of the discovery seems to be ok. Is there somewhere I can send the debug log output for the discover that isn't public?

**Comment by inphobia** [comment]
> Ok. Nothing in the default vrf and the mgmt0 port is on its own vrf.

while snmp::info::layer3::nexus tries to do it's best to enumerate vrfs, thus far it hasn't worked for me. with snmpv3 you should be able to use the `context_prefix` option (https://github.com/netdisco/netdisco/wiki/Configuration#device_auth), but i haven't had much luck with that either. having a look at that is on my todo list, but near the end. although with our recent network redesign it has started to move up.

i do seem to recall the mgmt0 interface showing up some time ago, perhaps https://github.com/netdisco/snmp-info/commit/604de81ab8ae0516e75ff34fb2b0ff142f5c9852#diff-798b6a50457fd18e5a93c9863fd7fa92 and/or https://github.com/netdisco/netdisco/commit/0bc5c68ed80e3551bab02a4a5d43415188a0fe05 and/or https://github.com/netdisco/netdisco/commit/3c16a0faaef1815a07f42df986361aeaba7a997 introduced a regression.

to be clear: with my current setup only the addresses in the vrf named "default" are showing up. #246 has a comment stating:
> On Nexus device, you need to add the following command for each VRF:
snmp-server context `<context>` vrf `<vrf>`

tried that but didn't work for me. even borked the snmp config on one of our nexuses trying to revert it, turns out you can delete implicitly deleted contexts...


as for the output, you can:
 * mail me: nick.nauwelaerts -at- aquafin.be
 * create a private github repo (they're free now), upload the output & provide us with access.
 * make the output available somewhere (perhaps a private github gist) & paste the link in our  irc channel: #netdisco on freenode.net

**Comment by ajw1980** [comment]
This seems to be a problem:
```
[23336] 2019-07-01 17:32:08 debug -> Can't use an undefined value as a HASH reference at /home/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Plugin/Discover/Properties.pm line 346.
[23336] 2019-07-01 17:32:08 debug Can't use an undefined value as a HASH reference at /home/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Plugin/Discover/Properties.pm line 346.
```

I commented out the ipv6 alias retrieval in Netdisco/Worker/Plugin/Discover/Properties.pm

`#  push @aliases, _get_ipv6_aliases($device, $snmp);`

and the IPv4 aliases got in the database.

**Comment by ollyg** [comment]
Thanks @ajw1980 this is fixed in 26960bdd347f34c03ab36eb052283f11975b5d70
